pkg/services/store/validate.go

```diff
@@ -96,6 +103,21 @@ func (s *standardStorageService) validateUploadRequest(ctx context.Context, user
 	case EntityTypeImage:
 		return s.validateImage(ctx, user, req)
 	default:
+		if t, exists := types[et]; exists {
+			val, err := cuectx.JSONtoCUE("", req.Contents)
+			if err != nil {
+				// TODO: Use Go errors
+				return fail(err.Error())
+			}
+
+			_, err = t.CurrentSchema().Validate(val)
+			if err != nil {
+				// TODO: Use Go errors
+				return fail(err.Error())
+			}
+			return success()
+		}
+
```

Comment:
i think this is a case where we'd want to use [Mux](https://github.com/grafana/grafana/blob/main/pkg/framework/coremodel/helpers.go#L108), which will batch translation with validation.

(Note, that'll replaced with a considerably smoother generic system - grafana/thema#63 - once we've updated Grafana to go1.18)

Reply:
and your TODOs remind me that the official upstream muxers should have an error taxonomy for differentiating classes of errors - was it a bad JSON encoding? validation failure? translation error (which will hopefully eventually be impossible)?